File: go/base/models/circle.go
package models

import (
	"fmt"
	"math"
	. "shapes/base/commons"
	"strings"
)

type Circle struct {
	BaseShape   *BaseShape
	Displayable *Displayable
	radius      float64
	color       string
}

func (Circle) New(radius float64) *Circle {
	var dimOne = map[bool]float64{true: radius, false: 1}[radius > 0]

	baseShape := BaseShape{}.New(
		[]map[string]float64{
			{"radius": math.Max(radius, 0)},
			{"diameter": math.Max(radius, 0) * 2},
			{"curvature": 1 / math.Max(radius, 0)},
		},
	)

	return &Circle{
		BaseShape:   baseShape,
		Displayable: &Displayable{},
		radius:      dimOne,
		color:       "None",
	}
}

func (c *Circle) GetColor() string {
	return c.color
}

func (c *Circle) SetColor(color string) {
	c.color = color
}

func (c *Circle) PrintDescription() {
	fmt.Printf("Circle radius: %.2f\n", c.radius)
	fmt.Printf("Circle dimensions: %v\n", c.dimensionFormatter())
	fmt.Printf("Circle color: %s\n", c.color)
	fmt.Printf("Circle area: %.2f\n", c.Area())
	fmt.Printf("Circle perimeter: %.2f\n", c.Perimeter())
	fmt.Printf("Circle is displayable: %t\n", c.Displayable.ShouldDisplay())
	fmt.Printf("Circle is a shape: %t\n", c.BaseShape.ValidateShape())
}

func (c Circle) Area() float64 {
	return math.Pi * math.Pow(c.radius, 2)
}

func (c Circle) Perimeter() float64 {
	return 2 * math.Pi * c.radius
}

func (c *Circle) dimensionFormatter() string {
	var parts []string
	for _, dimension := range c.BaseShape.Dimensions {
		for key, value := range dimension {
			parts = append(parts, fmt.Sprintf("%s: %.2f", strings.ToUpper(key), value))
		}
	}

	return strings.Join(parts, " | ")
}

File: cplusplus/src/base/models/circle.cpp
#define _USE_MATH_DEFINES

#include <boost/algorithm/string.hpp>
#include <boost/format.hpp>
#include <cmath>
#include <print>
#include <shapes/base/models.hpp>
#include <sstream>

using namespace boost::algorithm;
using namespace std;
using frmt = boost::format;

namespace shapes::base::models {
Circle::Circle(double radius)
  : BaseShape({
        {{"radius", std::max(radius, 1.0)}},
        {{"diameter", std::max(radius, 1.0) * 2}},
        {{"curvature", 1.0 / std::max(radius, 1.0)}},
    }) {
    const double dimOne = (radius > 0) ? radius : 1;
    this->radius = dimOne;
}

auto Circle::getColor() const -> const std::string& {
    return color;
}

void Circle::setColor(const std::string& value) {
    color = value;
}

void Circle::printDescription() const {
    std::println("Circle radius: {}", this->radius);
    std::println("Circle dimensions: {}", this->formatDimensions());
    std::println("Circle color: {}", this->color);
    std::println("Circle area: {}", this->area());
    std::println("Circle perimeter: {}", this->perimeter());
    std::println("Circle is displayable: {}", this->shouldDisplay());
    std::println("Circle is a shape: {}", Circle::validateShape());
}

auto Circle::area() const -> double {
    return M_PI * std::pow(radius, 2);
}

auto Circle::perimeter() const -> double {
    return 2 * M_PI * radius;
}

auto Circle::formatDimensions() const -> string {
    vector<string> parts;
    for (const auto& dimensions : getDimensions()) {
        for (const auto& [key, value] : dimensions) {
            ostringstream entry;
            entry << to_upper_copy(key) << ": " << str(frmt("%.3f") % value);
            parts.push_back(entry.str());
        }
    }

    return join(parts, " | ");
}
} // namespace shapes::base::models
